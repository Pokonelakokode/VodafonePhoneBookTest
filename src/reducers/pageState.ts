import {IPageStateActions, PageStateActionTypes} from "../actions/pageState";
import {getHash} from "../utils";

export type PageState = {
    selectedUser: number | null,
    search: string,
    letterFilter: string
}

const initialState:PageState = {
    selectedUser: parseInt(getHash().user) || null,
    search: "",
    letterFilter: ""
};


export default function pageState(state = initialState, action: IPageStateActions) {
    switch (action.type) {
        case PageStateActionTypes.SET:
            return {...state,[action.key]:action.data};
        default:
            return state
    }
}

