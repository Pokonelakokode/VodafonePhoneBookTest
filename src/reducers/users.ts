import {IUserActions, UserActionTypes} from "../actions/users";
import { User } from "../types";
import {loadFromStorage} from "../utils";

export default function users(state = loadFromStorage(), action: IUserActions):User[] {
    switch (action.type) {
        case UserActionTypes.LOAD:
            return action.users;
        case UserActionTypes.SET:
            return state.map(user => user.id === action.user.id ? action.user : user);
        case UserActionTypes.ADD:
            return [...state,action.user];
        case UserActionTypes.DELETE:
            return state.filter(user => user.id !== action.id);
        default:
            return state
    }
}