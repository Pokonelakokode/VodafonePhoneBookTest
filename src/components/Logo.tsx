import * as React from "react";
import { bindActionCreators } from "redux";
import {selectUser} from "../actions/pageState";
import {useDispatch} from "react-redux";

interface IProps {
    imgSrc: string
}

const Logo: React.FC<IProps> = ({imgSrc}) => {
    const actions = bindActionCreators({selectUser},useDispatch());
    return (
        <div className="col-12 col-md-6 pointer-cursor">
            <img onClick={() => actions.selectUser(null)} className="logo" src={imgSrc} />
        </div>
    )
};

export default Logo