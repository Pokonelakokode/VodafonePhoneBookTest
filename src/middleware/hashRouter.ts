import {Dispatch, Middleware} from "redux";
import {RootActions} from "../actions";
import {PageStateActionTypes} from "../actions/pageState";
import {setHash} from "../utils";

export const hashRouter:Middleware = () => (next:Dispatch) => (action:RootActions) => {
    next(action);
    if(action.type === PageStateActionTypes.SET && action.key === "selectedUser"){
        action.data ? setHash({user:action.data}) : setHash();
    }
};