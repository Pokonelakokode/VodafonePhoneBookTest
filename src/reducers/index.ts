import { combineReducers } from "redux";
import pageState from "./pageState";
import users from "./users";

export const rootReducer = combineReducers({
    pageState,
    users
});

export type RootState = ReturnType<typeof rootReducer>