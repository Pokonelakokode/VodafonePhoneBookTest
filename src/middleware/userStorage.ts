import {Dispatch, Middleware, MiddlewareAPI} from "redux";
import {RootActions} from "../actions";
import {UserActionTypes} from "../actions/users";
import {saveToStorage} from "../utils";
import {RootState} from "../reducers";

export const userStorage:Middleware = (store: MiddlewareAPI<any,RootState>) => (next:Dispatch) => (action:RootActions) => {
    const {DELETE, ADD, SET} = UserActionTypes;
    next(action);
    if(action.type === DELETE || ADD || SET){
        saveToStorage(store.getState().users)
    }
};

export default userStorage