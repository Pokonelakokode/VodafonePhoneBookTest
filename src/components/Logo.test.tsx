import * as React from "react";
import * as renderer from "react-test-renderer"
import {createStore, applyMiddleware} from "redux";
import {rootReducer} from "../reducers";
import {hashRouter} from "../middleware/hashRouter";
import userStorage from "../middleware/userStorage";
import { Provider } from "react-redux";
import Logo from "./Logo";

describe("Logo",() => {
    const store = createStore(rootReducer, applyMiddleware(hashRouter, userStorage));
    it('should render snapshot', function () {
        const component = renderer.create(
            <Provider store={store}>
                <Logo imgSrc="vodafone_logo.svg"/>
            </Provider>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
});