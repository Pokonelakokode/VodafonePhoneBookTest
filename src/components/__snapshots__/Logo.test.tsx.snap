// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logo should render snapshot 1`] = `
<div
  className="col-12 col-md-6 pointer-cursor"
>
  <img
    className="logo"
    onClick={[Function]}
    src="vodafone_logo.svg"
  />
</div>
`;
